logstash-core/src/main/java/org/logstash/instrument/metrics/HistoricMetric.java

```diff
@@ -0,0 +1,230 @@
+
+    // update- and retrieve-order quick reference lists.
+    private final List<QuickReference> quickUpdateReferences = List.of(quickLifetime, quickInstant, quickOneMinute, quickFiveMinutes, quickFifteenMinutes);
+    private final List<QuickReference> quickReadReferences =   List.of(quickInstant, quickOneMinute, quickFiveMinutes, quickFifteenMinutes, quickLifetime);
+
+    // Safeguard for reads against a dead writer
+    private final AtomicReference<Instant> lastCaptured = new AtomicReference<>();
+    private static final Duration STALENESS_THRESHOLD = Duration.ofSeconds(10);
+
+    public HistoricMetric(Metric<T> numeratorMetric, Metric<Long> denominatorMetric) {
+        this(Clock.systemUTC(), numeratorMetric, denominatorMetric);
+    }
+
+    public HistoricMetric(Metric<T> metric) {
+        this(Clock.systemUTC(), metric);
+    }
+
+    HistoricMetric(Clock clock, Metric<T> numeratorMetric, Metric<Long> denominatorMetric) {
+        this.clock = Objects.requireNonNull(clock, "clock must not be null");
+        this.numeratorMetric = Objects.requireNonNull(numeratorMetric, "numeratorMetric must not be null");
+        this.denominatorMetric = Objects.requireNonNullElseGet(denominatorMetric, () -> new UptimeMetric(clock, null));
+    }
+
+    HistoricMetric(Clock clock, Metric<T> metric) {
+        this(clock, metric, null);
+    }
+
+    /**
+     * Capture the current values of the linked metrics
+     * TODO: capturing of an HistoricMetric will need to be scheduled on a repeating cadence
+     */
+    public void capture() {
+        capture(clock.instant(), numeratorMetric.getValue(), denominatorMetric.getValue());
+    }
+
+    void capture(final Instant timestamp, final T numerator, final Long denominator) {
+        final Capture capture = new Capture(timestamp, numerator, denominator);
+        final Capture previousTail = this.tail.getAndSet(capture);
+        previousTail.next.set(capture);
+
+        this.lastCaptured.set(timestamp);
+
+        // rotate our quick references in ascending age order,
+        // bailing entirely if we are too young to warrant continuing
+        for (QuickReference quickReference : quickUpdateReferences) {
+            if (quickReference.rotate(timestamp) == false) { return; }
+        }
+
+        // trim out elements that are no longer necessary
+        compact();
+    }
+
+    // We do not need entries between our earliest capture and the one
+    // referenced by our 15-minutes-ago quick reference, so we compact
+    // our linked list to free the nodes in between.
+    private void compact() {
+        final Capture earliest_capture = quickLifetime.capture.get();
+        if (Objects.isNull(earliest_capture)) { return; }
+
+        final Capture fifteen_minutes_ago = quickFifteenMinutes.capture.get();
+        if (Objects.isNull(fifteen_minutes_ago)) { return; }
+
+        earliest_capture.next.set(fifteen_minutes_ago);
```

Comment:
Can we keep camel casing for vars?

```suggestion
        final Capture earliestCapture = quickLifetime.capture.get();
        if (Objects.isNull(earliestCapture)) { return; }

        final Capture fifteenMinutesAgo = quickFifteenMinutes.capture.get();
        if (Objects.isNull(fifteenMinutesAgo)) { return; }

        earliestCapture.next.set(fifteenMinutesAgo);
```